src/Relation/Binary/Rewriting.agda

```diff
@@ -0,0 +1,135 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Concepts from rewriting theory
+-- Definitions are based on "Term Rewriting Systems" by J.W. Klop
+------------------------------------------------------------------------
+
+{-# OPTIONS --with-K --safe #-}
+
+module Relation.Binary.Rewriting where
+
+open import Agda.Builtin.Equality using (_≡_ ; refl)
+open import Data.Product using (_×_ ; ∃ ; _,_ ; proj₂)
+open import Data.Empty
+open import Data.Sum as Sum using (_⊎_)
+open import Level
+open import Relation.Binary.Core
+open import Relation.Binary.Construct.Closure.Transitive
+open import Relation.Binary.Construct.Closure.ReflexiveTransitive
+open import Relation.Binary.Construct.Closure.Symmetric
+open import Relation.Nullary
+
+-- The following definitions are taken from Klop [5]
+IsNormalForm : ∀ {A : Set} → {ℓ : Level} → {r : Rel A ℓ} → (a : A) → Set _
+IsNormalForm {A} {ℓ} {_⟶_} a = ¬ ∃ λ b → (a ⟶ b)
+
+HasNormalForm : ∀ {A : Set} → {ℓ : Level} → {r : Rel A ℓ} → (b : A) → Set _
+HasNormalForm {A} {ℓ} {_⟶_} b = ∃ λ a → ( IsNormalForm {A} {ℓ} {_⟶_} a × (b —↠ a))
+  where
+    _—↠_ = Star _⟶_
+
+NormalForm : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+NormalForm {A} {ℓ} r = ∀ {a b}
+  → IsNormalForm {A} {ℓ} {r} a
+  → b ↔ a
+  → b —↠ a
+  where
+    _—↠_ = Star r
+    _↔_  = Star (SymClosure r)
+
+WeaklyNormalizing : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+WeaklyNormalizing {A} {ℓ} r = ∀ a → HasNormalForm {A} {ℓ} {r} a
+
+UniqueNormalForm : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+UniqueNormalForm {A} {ℓ} r = ∀ {a b}
+  → IsNormalForm {A} {ℓ} {r} a
+  → IsNormalForm {A} {ℓ} {r} b
+  → a ↔ b
+  → a ≡ b
+  where
+    _↔_ = Star (SymClosure r)
+
+Det : ∀ {a b ℓ₁ ℓ₂} → {A : Set a} → {B : Set b} → Rel B ℓ₁ → REL A B ℓ₂ → Set _
+Det _≈_ _—→_ = ∀ {x y z} → x —→ y → x —→ z → y ≈ z
+
+Deterministic : ∀ {a b ℓ} → {A : Set a} → {B : Set b} → REL A B ℓ → Set _
+Deterministic = Det _≡_
+
+Confluent : ∀ {A : Set} → {ℓ : Level} → (r :  Rel A ℓ) → Set _
+Confluent _—→_ = ∀ {A B C} → (A —↠ B × A —↠ C) → ∃ λ D → (B —↠ D) × (C —↠ D)
+  where
+     _—↠_ = Star _—→_
+
+Diamond : ∀ {A : Set} → {ℓ : Level} → (r :  Rel A ℓ) → Set _
+Diamond _—→_ = ∀ {A B C} → (A —→ B × A —→ C) → ∃ λ D → (B —↠ D) × (C —↠ D)
+  where
+    _—↠_ = Star _—→_
+
+det⟶conf : ∀ {A : Set}
+  → {ℓ : Level}
```

Comment:
By convention we put the arrows at the end of the line rather than the beginning of the next. I've just updated the [style guide](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md) with a note.

If you use the style mentioned above (`∀ {a ℓ} {A : Set a} → ... etc.) this should all fit on one line anyway.

Also by convention we capitalise the relations (i.e. `R : Rel A ℓ`) but in actual fact in this case it would be better if you used the same symbol as used in the definition of the type (i.e. `_—→_ : Rel A ℓ`)